### PR TITLE
[Resource][Data Source] ACR Scope Map and Tokens

### LINKED ISSUE #6505: Support for Azure Container Registry tokens and scope
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

ACR supports token authentication that allows to fine-tune the permissions associated to the token: push, pull, delete, list.

### New or Affected Resource(s)

* azurerm_container_registry
* azurerm_container_registry_token
* azurerm_container_registry_scope_map

### Potential Terraform Configuration

```hcl
resource "azurerm_container_registry" "acr" {
  name                     = "containerRegistry1"
  resource_group_name      = azurerm_resource_group.rg.name
  location                 = azurerm_resource_group.rg.location
  sku                      = "Premium"
  admin_enabled            = false
  georeplication_locations = ["East US", "West Europe"]
}

resource "azurerm_container_registry_token" "acr-pull" {
  name                  = "pull-image"
  container_registry_id = azurerm_container_registry.acr.id
  scope_map_id          = data.azurerm_container_registry_scope_map.pull-only
}

# ACR includes some default system scope maps
data "azurerm_container_registry_scope_map" "pull-only" {
  name                  = "_repositories_pull"
  container_registry_id = azurerm_container_registry.acr.id
  type                  = "System"
}

```

### References

* https://docs.microsoft.com/en-ca/azure/container-registry/container-registry-repository-scoped-permissions
* https://docs.microsoft.com/en-us/cli/azure/acr/scope-map?view=azure-cli-latest
* https://docs.microsoft.com/en-us/cli/azure/acr/token?view=azure-cli-latest

### PR DESCRIPTION
**Pull request by MattiasAng**
Supersede #9402 after email exchange with @Lucretius which didn't have time to complete it currently.
I fixed the comments from last review.

Data Source for azurerm_container_registry_token
Data Source for azurerm_container_registry_scope_map
Resource for azurerm_container_registry_token
Resource for azurerm_container_registry_scope_map
Documentation for each
Tests for each, they all pass running against my own Azure environment.
There are passwords for the token which I would like to expose but there is no way to delete them once generated so they didn't really fit into the lifecycle management of the token itself. Perhaps they might be better suited as a separate data source, similar to how we handle the SAS tokens?

Also note that these are still both preview features. As such I've vendor-included the SDK preview code and only used it in these specific preview features (see client.go)

This PR resolves #6505

### CODE REVIEW

File: azurerm/internal/services/containers/container_registry_scope_map_data_source.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_scope_map_data_source.go:66]: can we make sure ID is not nil here with a nnil check?

File: azurerm/internal/services/containers/container_registry_scope_map_data_source.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_scope_map_data_source.go:26]: since we have a validation function for the name can we use it here to?

File: azurerm/internal/services/containers/container_registry_scope_map_resource.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_scope_map_resource.go:180]: could we use the id var instead?
```suggestion

	resp, err := client.Get(ctx, id.ResourceGroup, id.ContainerRegistryName, id.Name)
```

File: azurerm/internal/services/containers/container_registry_scope_map_resource.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_scope_map_resource.go:214]: here too
```suggestion

	future, err := client.Delete(ctx, id.ResourceGroup, id.ContainerRegistryName, id.Name)
```

File: azurerm/internal/services/containers/container_registry_scope_map_resource_test.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_scope_map_resource_test.go:72]:
```suggestion
func TestAccontainerRegistryScopeMap_update(t *testing.T) {
```

File: azurerm/internal/services/containers/container_registry_scope_map_resource_test.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_scope_map_resource_test.go:84]: could we put an import step here

File: azurerm/internal/services/containers/container_registry_scope_map_resource_test.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_scope_map_resource_test.go:93]: and another here

File: azurerm/internal/services/containers/container_registry_scope_map_resource_test.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_scope_map_resource_test.go:98]: can we use the generated id parser here?

File: azurerm/internal/services/containers/container_registry_token_data_source.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_token_data_source.go:26]: can we validate this?

File: azurerm/internal/services/containers/container_registry_token_data_source.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_token_data_source.go:63]: could we make sure this isn't nil with a nil check?

File: website/docs/r/container_registry_token.html.markdown
Comment (katbyte) [inline website/docs/r/container_registry_token.html.markdown:60]: could we just name this enabled and make it a boolean?

File: website/docs/r/container_registry_scope_map.html.markdown
Comment (katbyte) [inline website/docs/r/container_registry_scope_map.html.markdown:51]:
```suggestion
* `actions` - (Required) A list of actions to attach to the scope map (e.g. `repo/content/read`, `repo2/content/delete`).
```

File: azurerm/internal/services/containers/container_registry_token_resource_test.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_token_resource_test.go:100]: could we use the generated parse function here to?
```suggestion
	id, err := azure.ParseACR(state.ID)
	if err != nil {
		return nil, err
	}
	
	resp, err := clients.Containers.TokensClient.Get(ctx, id.ResourceGroup, id.RegistryName, id.Name)
```

File: azurerm/internal/services/containers/container_registry_token_resource.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_token_resource.go:215]: same here
```suggestion

	future, err := client.Delete(ctx, id.ResourceGroup, id.ContainerRegistryName, id.Name)
```

File: azurerm/internal/services/containers/container_registry_token_resource.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_token_resource.go:181]: and here

File: azurerm/internal/services/containers/container_registry_scope_map_data_source.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_scope_map_data_source.go:37]: could we add some whitespace here to be consistent with the majority of the provider?
```suggestion
			"container_registry_name": {
				Type:         schema.TypeString,
				Required:     true,
				ValidateFunc: ValidateContainerRegistryName,
			},
			
			"resource_group_name": azure.SchemaResourceGroupNameForDataSource(),
			
			"description": {
				Type:     schema.TypeString,
				Computed: true,
			},
			
			"actions": {
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for taking on this pr @MattiasAng - overall this looks great and most of my comments are minor. this should be good to merge once they have been addressed

File: website/docs/r/container_registry_token.html.markdown
Comment (MattiasAng) [reply website/docs/r/container_registry_token.html.markdown:60]: `azurerm_container_registry_webhook` uses the same syntax so I'll let you decide if we want to keep it the same over all sub-resources of Container Registry or keep it like it is.

File: azurerm/internal/services/containers/container_registry_scope_map_data_source.go
Comment (MattiasAng) [reply azurerm/internal/services/containers/container_registry_scope_map_data_source.go:26]: There are system scope maps that uses a different naming standard so we cannot use the same validation for the data source.

File: website/docs/r/container_registry_token.html.markdown
Comment (katbyte) [reply website/docs/r/container_registry_token.html.markdown:60]: Let's change it to enabled please, status = enabled was the older way of doing things. enabled = true is far more clear and autocompletes better. If more status are added we can always deprecate this property

**Comment by katbyte** [comment]
see my comment left inline @MattiasAng

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks, this should be good to merge once the build is fix:
```
Run GOARCH=386 GOOS=linux go build -o 32bitbuild .
# github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers
azurerm/internal/services/containers/container_registry_scope_map_data_source.go:30:19: undefined: ValidateContainerRegistryName
azurerm/internal/services/containers/container_registry_scope_map_resource.go:42:19: undefined: ValidateContainerRegistryName
azurerm/internal/services/containers/container_registry_scope_map_resource.go:57:19: undefined: ValidateContainerRegistryName
azurerm/internal/services/containers/container_registry_token_data_source.go:27:19: undefined: ValidateContainerRegistryName
azurerm/internal/services/containers/container_registry_token_data_source.go:33:19: undefined: ValidateContainerRegistryName
azurerm/internal/services/containers/container_registry_token_resource.go:42:19: undefined: ValidateContainerRegistryName
azurerm/internal/services/containers/container_registry_token_resource.go:51:19: undefined: ValidateContainerRegistryName
Error: Process completed with exit code 2.
```

**Comment by MattiasAng** [comment]
Not sure why it is failing. I have no problem building locally (Windows and WSL2) and ValidateContainerRegistryName is still defined.

File: azurerm/internal/services/containers/container_registry_scope_map_data_source_test.go
Comment (katbyte) [inline azurerm/internal/services/containers/container_registry_scope_map_data_source_test.go:26]:
```suggestion
```
this doesn't seem to exist in schema

**Review by katbyte — APPROVED** [review]
Thanks @MattiasAng - aside from 2 comments this looks good 👍

**Pull request merged**
